### ISSUE TITLE
0.31 fails to compile

### ISSUE THREAD

**Issue by bestouff** [AtherEnergy/rumqtt#174]
```bash
   Compiling rumqtt v0.31.0
error[E0308]: mismatched types
   --> /home/xav/.cargo/registry/src/github.com-1ecc6299db9ec823/rumqtt-0.31.0/src/client/network.rs:130:34
    |
130 |             config.set_protocols(&self.alpn_protocols);
    |                                  ^^^^^^^^^^^^^^^^^^^^ expected slice, found struct `std::vec::Vec`
    |
    = note: expected type `&[std::string::String]`
               found type `&std::vec::Vec<std::vec::Vec<u8>>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `rumqtt`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by TheBestJohn** [comment]
What other crates are you using in this? This is due to an old rustls version being used. If you're just using this crate, do a cargo clean.

**Comment by bestouff** [comment]
I'm using it in a project with many other crates; maybe the problem is that the rustls dependency isn't correctly specified.

**Comment by tekjar** [comment]
@bestouff Can you try removing Cargo.lock and target dir and then recompile?

**Comment by TheBestJohn** [comment]
Yeah anything that depends on an older version of ring will fail with this. rustls used to use `String` for alpn_protocols [docs](https://docs.rs/rustls/0.14.0/rustls/struct.ClientConfig.html#structfield.alpn_protocols) and in version 0.15.0 onwards used `Vec<u8>` [docs](https://docs.rs/rustls/0.15.0/rustls/struct.ClientConfig.html#structfield.alpn_protocols)

**Comment by bestouff** [comment]
I tried recompiling without `Cargo.lock` and effectively `rustls` went from 0.14 to 0.15, and my project compiled. `tokio-rustls` also got a bump from 0.8 to 0.9, which is where `rustls` is pulled from.  
So the problem lies in `Cargo.toml`'s deps. I'm sending a patch.

**Comment by bestouff** [comment]
See MR #175

**Comment by tekjar** [comment]
Did you try removing both target and Cargo.lock?

I'm ok with the pull request but I think there are people using rumqtt with other dependencies which require older version of ring. Wouldn't pinning tokio-rustls pin compatible ring as well causing conflicts with other dependencies depending on older ring?

**Comment by bestouff** [comment]
> Did you try removing both target and Cargo.lock at a time?  

Yes I did.  

> Wouldn't pinning tokio-rustls pin compatible ring as well causing conflicts with other dependencies depending on older ring?  

I don't know. But now it may not compile, so there's obviously a bug.

**Comment by tekjar** [comment]
@TheBestJohn Are you able to compile 0.31?

**Comment by bestouff** [comment]
I think only people with a dependency on `rustls` 0.14 will see a problem when creating a new project.

**Comment by TheBestJohn** [comment]
> @TheBestJohn Are you able to compile 0.31?

I can compile it by itself just fine but the alpn stuff gets in the way if I'm using it with Rocket.

**Comment by PetrGlad** [comment]
Also Rumqtt version 0.31 has no tag - it is hard to see which sources were actually used.

**Comment by PetrGlad** [comment]
I cannot build project using rumqtt 0.31.0, with same error. I, however can `cargo build` this project separately from master.
Note that published crate on crates.io uses tokio-rustls 0.8.1 (which in turn requires rustls 0.14)
while project from git /master here requires tokio-rustls 0.9.3 (which in turn requires rustls 0.15.1).
Actual versions are from `Cargo.lock`
So, yes it would definitely help if each released version would have a tag in git here.

Also I am not sure what is expected behavior here but rumqtt's `Cargo.toml` has requirement
`tokio-rustls = ">=0.8, <=0.9"` while actual used version is `tokio-rustls 0.9.3`

**Comment by PetrGlad** [comment]
The fix would be to explicitly require tokio-rustls version  0.9.1 or later.

**Comment by TheBestJohn** [comment]
I don't think we can get away from historical rustls and ring issues here. It's a pain of moving forward. I also don't think it's incumbent on a libraries developer to support other libraries using old sources.

**Comment by feymartynov** [comment]
Had this error on macOS after system upgrade from Mojave to Catalina. Removing `Cargo.lock` and `target` dir of the project solved the problem.

**Comment by TotalKrill** [comment]
Which is just lucky and depending on other deps in the system, this project is now dependent on tokio-rustls ^9 and therefore the cargo.toml is essentially wrong for now
